__tests__/lib/messageUtils.test.ts

```diff
@@ -81,7 +81,7 @@ describe("getDashboard", () => {
       branchSlug,
     );
 
-    const expectedLink = `https://mozilla.cloud.looker.com/dashboards/${dashboardId}?Messaging+System+Ping+Type=${encodeURIComponent(template)}&Submission+Date=30%20day%20ago%20for%2030%20day&Messaging+System+Message+Id=${encodeURIComponent(msgId)}&Normalized+Channel=${encodeURIComponent(channel)}&Normalized+OS=&Client+Info+App+Display+Version=&Normalized+Country+Code=&Experiment=${encodeURIComponent(experiment)}&Experiment+Branch=${encodeURIComponent(branchSlug)}`;
+    const expectedLink = `https://mozilla.cloud.looker.com/dashboards/${dashboardId}?Messaging+System+Ping+Type=${encodeURIComponent(template)}&Submission+Date=30%20day%20ago%20for%2030%20day&Messaging+System+Message+Id=${encodeURIComponent(msgId)}%2C+${encodeURIComponent(msgId).toLowerCase()}%2C+${encodeURIComponent(msgId).toUpperCase()}&Normalized+Channel=${encodeURIComponent(channel)}&Normalized+OS=&Client+Info+App+Display+Version=&Normalized+Country+Code=&Experiment=${encodeURIComponent(experiment)}&Experiment+Branch=${encodeURIComponent(branchSlug)}`;
```

Comment:
One of the things that I've been struggling with is that it feels like we've hit a tipping point where it's now hard to have confidence that, especially for tests, but also to some degree the code under test, it's hard to have confidence that:

a) the current state is actually correct
b) changes to the current state aren't breaking something

I think that a big part of what's going on here is that we're doing all the encoding and testing of the encoding by hand, at the same time as we're also trying to test that they are creating URLs that do what we want to do here.

Which is another way of saying that I think some separation of concerns would help meaningfully.  More in upcoming comments...

Reply:
Would love hear thoughts from you and @AllegroFox

Reply:
Yeah, this makes sense. I tried updating the tests in `messageUtils.test.ts` so we don't have to write the encoding and the full URLs by hand. Instead, I'm checking the URLSearchParams to check the values for certain parameters which will also do the encoding for me.

Reply:
I'm just not entirely sure how to update the tests in `nimbusRecipe.test.ts` similarly because the tests aren't specifically testing the dashboard links but are failing because they're included in the `branchInfo` objects.

Reply:
How about something like this:

```diff
diff --git a/__tests__/lib/nimbusRecipe.test.ts b/__tests__/lib/nimbusRecipe.test.ts
index aded435..bdc359a 100644
--- a/__tests__/lib/nimbusRecipe.test.ts
+++ b/__tests__/lib/nimbusRecipe.test.ts
@@ -1,7 +1,9 @@
 import { NimbusRecipe } from "@/lib/nimbusRecipe";
 import { ExperimentFakes } from "@/__tests__/ExperimentFakes.mjs";
 import { BranchInfo } from "@/app/columns.jsx";
-import { getDashboardIdForTemplate } from "@/lib/messageUtils";
+import { formatDate } from "@/lib/experimentUtils";
+import { getExperimentLookerDashboardDate } from "@/lib/lookerUtils";
+import { getDashboard, getDashboardIdForTemplate } from "@/lib/messageUtils";
 
 //XXX We're passing this custom object for about:welcome, since ExperimentFakes
 // doesn't quite give us what we want in that case. We probably want to tweak
@@ -136,14 +138,35 @@ describe("NimbusRecipe", () => {
       const branch = AW_RECIPE.branches[1];
 
       const branchInfo = nimbusRecipe.getBranchInfo(branch);
+      const proposedEndDate = getExperimentLookerDashboardDate(
+        branchInfo.nimbusExperiment.startDate,
+        branchInfo.nimbusExperiment.proposedDuration,
+      );
       const dashboardId = getDashboardIdForTemplate("aboutwelcome");
+      const formattedEndDate = formatDate(
+        branchInfo.nimbusExperiment.endDate as string,
+        1,
+      );
+
+      const dashboardLink = getDashboard(
+        branch.template as string, // XXX this seems wrong
+        branchInfo.id,
+        undefined,
+        branchInfo.nimbusExperiment.slug,
+        branch.slug,
+        branchInfo.nimbusExperiment.startDate,
+        branchInfo.nimbusExperiment.endDate
+          ? formattedEndDate
+          : proposedEndDate,
+        false, // default for isCompleted in constructor
+      );
 
       // XXX getBranchInfo is actually going to return a previewLink, which
       // makes this test kind of brittle. We could refactor this to no longer
       // use deepEqual and check for the existence of object properties instead.
       expect(branchInfo).toEqual({
         product: "Desktop",
-        ctrDashboardLink: `https://mozilla.cloud.looker.com/dashboards/${dashboardId}?Submission+Timestamp+Date=30%20day%20ago%20for%2030%20day&Message+ID=%25feature_value_id%3Atreatment-a%25%2C+%25${"feature_value_id%3Atreatment-a".toLowerCase()}%25%2C+%25${"feature_value_id%3Atreatment-a".toUpperCase()}%25&Normalized+Channel=&Experiment=aboutwelcome-test-recipe&Branch=treatment-a`,
+        ctrDashboardLink: dashboardLink,
         id: "feature_value_id:treatment-a",
         isBranch: true,
         nimbusExperiment: AW_RECIPE,
```